staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -143,107 +137,168 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
 			return
 		}
 
-		// pass request to the next handler if found the gvr locally.
-		// TODO: maintain locally serviceable GVRs somewhere so that we dont have to
-		// consult the storageversion-informed map for those
-		if locallyServiceable {
-			handler.ServeHTTP(w, r)
-			return
+		rand := rand.Intn(len(peerEndpoints))
+		peerEndpoint := peerEndpoints[rand]
+		h.proxyRequestToDestinationAPIServer(r, w, peerEndpoint)
+	})
+}
+
+func (h *peerProxyHandler) startLocalDiscoveryCacheInvalidation(stopCh <-chan struct{}) {
+	go func() {
+		klog.Info("localDiscoveryCacheInvalidation goroutine started")
+		for {
+			select {
+			case <-h.localDiscoveryCacheTicker.C:
+				klog.V(4).Infof("Invalidating local discovery cache")
+				if err := h.populateLocalDiscoveryCache(); err != nil {
+					klog.Errorf("Failed to repopulate local discovery cache: %v", err)
+				}
+			case <-stopCh:
+				klog.Info("localDiscoveryCacheInvalidation goroutine received stop signal")
+				if h.localDiscoveryCacheTicker != nil {
+					h.localDiscoveryCacheTicker.Stop()
+					klog.Info("localDiscoveryCacheTicker stopped")
+				}
+				klog.Info("localDiscoveryCacheInvalidation goroutine exiting")
+				return
+			}
 		}
+	}()
+}
 
-		if len(peerEndpoints) == 0 {
-			klog.Errorf("gvr %v is not served by anything in this cluster", gvr)
-			handler.ServeHTTP(w, r)
-			return
+func (h *peerProxyHandler) populateLocalDiscoveryCache() error {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(h.loopbackClientConfig)
+	if err != nil {
+		return fmt.Errorf("error creating discovery client: %w", err)
+	}
+
+	_, resourcesByGV, _, err := discoveryClient.GroupsAndMaybeResources()
+	if err != nil {
+		return fmt.Errorf("error getting API group resources from discovery: %w", err)
+	}
+
+	freshLocalDiscoveryResponse := map[schema.GroupVersion][]string{}
+	for gv, resources := range resourcesByGV {
+		if gv.Group == "" {
+			gv.Group = "core"
 		}
+		freshLocalDiscoveryResponse[gv] = []string{}
+		for _, resource := range resources.APIResources {
+			freshLocalDiscoveryResponse[gv] = append(freshLocalDiscoveryResponse[gv], resource.Name)
+		}
+	}
 
-		// otherwise, randomly select an apiserver and proxy request to it
-		rand := rand.Intn(len(peerEndpoints))
-		destServerHostPort := peerEndpoints[rand]
-		h.proxyRequestToDestinationAPIServer(r, w, destServerHostPort)
-	})
+	h.localDiscoveryResponseCache = freshLocalDiscoveryResponse
+	return nil
 }
 
-func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource) (*sync.Map, error) {
-	apiserversi, ok := h.svMap.Load(gvr)
-	if !ok || apiserversi == nil {
-		return nil, fmt.Errorf("no storageVersions found for the GVR: %v", gvr)
+func (h *peerProxyHandler) shouldServeLocally(gvr schema.GroupVersionResource) bool {
+	resources, ok := h.localDiscoveryResponseCache[gvr.GroupVersion()]
```

Comment:
Why don't we just make the localDiscoveryResponseCache key a gvr instead of gv? Then you can do a direct map lookup instead of having to iterate.

Reply:
Another miss on my end, done now. Thanks!